sql/sql_select.cc

```diff
@@ -27147,15 +27147,48 @@ static bool setup_sum_funcs(THD *thd, Item_sum **func_ptr)
 }
 
 
-static bool prepare_sum_aggregators(THD *thd,Item_sum **func_ptr,
+bool JOIN::prepare_sum_aggregators(THD *thd,Item_sum **func_ptr,
                                     bool need_distinct)
 {
   Item_sum *func;
   DBUG_ENTER("prepare_sum_aggregators");
   while ((func= *(func_ptr++)))
   {
+    bool need_distinct_aggregator= need_distinct && func->has_with_distinct();
+    if (need_distinct_aggregator && table_count - const_tables == 1)
+    {
+      /*
+        Check if we can optimize aggregation and avoid de-duplication
+        of values. If there is:
+        - only one table involved in the join
+        - all arguments of the aggregate function are fields (not functions)
+        - there is a unique index on all those fields,
+        then the values retrieved are guaranteed to be unique, so there's
+        no need to do the de-duplication and we can employ SIMPLE_AGGREGATOR
+      */
+      List<Item> fields_list;
+      bool all_args_are_fields= true;
+      for (uint i= 0; i < func->argument_count(); i++)
+      {
+        if (func->arguments()[i]->type() == Item::FIELD_ITEM)
+        {
+          fields_list.push_back(func->arguments()[i]);
+        }
+        else
+        {
+          all_args_are_fields= false;
+          break;
+        }
+      }
+
+      if (all_args_are_fields &&
+          list_contains_unique_index(join_tab[const_tables].table,
+                                     find_field_in_item_list,
+                                     (void *) &fields_list))
+        need_distinct_aggregator= false;
+    }
     if (func->set_aggregator(thd,
-                             need_distinct && func->has_with_distinct() ?
+                             need_distinct_aggregator ?
```

Comment:
Pity there's no ANALYZE FORMAT=JSON or EXPLAIN output to include which aggregator is used. If its stable to add, it would be nice.

Reply:
Good idea, added logging to the optimizer trace. Now we can test not only the correct results but the fact the optimization is applied